purchase_invoice.py

```diff
@@ -875,7 +884,188 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
+                         default=paid_by_submitter, k='-paid-')],
+            [sg.Text('Kommentar')],
+            [sg.Input(k='-remarks-')],
+            [sg.Button('Speichern')]
+        ]
+        window1 = sg.Window("Einkaufsrechnung", layout, finalize=True)
+        window1.bring_to_front()
+        event, values = window1.read()
+        window1.close()
+        if values:
+            if '-supplier-' in values:
+                self.supplier = values['-supplier-']
+                if self.supplier == 'neu' and '-supplier-name-' in values:
+                    self.supplier = values['-supplier-name-']
+            if '-no-' in values:
+                self.no = values['-no-']
+            if '-date-' in values:
+                date = utils.convert_date4(values['-date-'])
+                if date:
+                    self.date = date
+            if '-vat-' in values:
+                self.vat[self.default_vat] = utils.read_float(values['-vat-'])
+            if '-gross-' in values:
+                self.totals[self.default_vat] = utils.read_float(values['-gross-']) - self.vat[self.default_vat]
+            if '-skonto-' in values and values['-skonto-']:
+                self.skonto = utils.read_float(values['-skonto-'])
+            if '-paid-' in values and values['-paid-']:
+                self.paid_by_submitter = True
+            if '-remarks-' in values:
+                self.remarks = values['-remarks-']
+        else:
+            return None
+        self.compute_total()
+        account = None
+        accounts = self.company.leaf_accounts_for_credit
+        account_names = [acc['name'] for acc in accounts]
+        if default_account:
+            for acc in account_names:
+                if default_account in acc:
+                    account = acc
+        if not account:
+            pinvs = self.company.purchase_invoices[self.supplier]
+            paccs = [pi['expense_account'] for pi in pinvs if 'expense_account' in pi]
+            paccs = list(set(paccs))
+            for acc in paccs:
+                try:
+                    account_names.remove(acc)
+                except Exception:
+                    pass
+            account_names = paccs + account_names
+            title = 'Buchungskonto wählen'
+            msg = 'Bitte ein Buchungskonto wählen\n'
+            account = easygui.choicebox(msg, title, account_names)
+            if not account:
+                return None
+        self.assign_default_e_items({self.default_vat: account})
+        return self
+
+    def parse_invoice_json(self, invoice_json, account=None, paid_by_submitter=False, given_supplier=None, is_test=False):
+        self.extract_items = self.update_stock
+        supps = dict(PurchaseInvoice.suppliers)
+        if self.company.name != 'Laden':
+            del supps['Rechnung']
+        try:
+            self.supplier = [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'supplier_name'][0]
```

Comment:
for PreR00492, I get `TypeError: string indices must be integers` here

Reply:
Which command was used? I need to replicate this error

Reply:
I have done some refactoring. Running play.py shows the error.

Reply:
By the way, if there is no supplier, you can take the one given in the arguments of the method,

Reply:
Is fixed now, see d2d170a083291793b9a5804fa039036ebaa5d4d2